src/hotspot/share/utilities/bitMap.hpp

```diff
@@ -354,84 +315,115 @@ class BitMap {
 #endif
 };
 
-// A concrete implementation of the "abstract" BitMap class.
+// CRTP: BitmapWithAllocator exposes to the following Allocator interfaces upward to GrowableBitMap.
 //
-// The BitMapView is used when the backing storage is managed externally.
-class BitMapView : public BitMap {
- public:
-  BitMapView() : BitMap(NULL, 0) {}
-  BitMapView(bm_word_t* map, idx_t size_in_bits) : BitMap(map, size_in_bits) {}
-};
+//  bm_word_t* allocate(idx_t size_in_words) const;
+//  void free(bm_word_t* map, idx_t size_in_words) const
+//
+template <class BitMapWithAllocator>
+class GrowableBitMap : public BitMap {
+ protected:
+  bm_word_t* reallocate(bm_word_t* map, idx_t old_size_in_bits, idx_t new_size_in_bits, bool clear = true);
 
-// A BitMap with storage in a ResourceArea.
-class ResourceBitMap : public BitMap {
+  GrowableBitMap(idx_t size_in_bits, bool clear) : BitMap(reallocate(nullptr, 0, size_in_bits, clear), size_in_bits) {
+  }
+  GrowableBitMap() : BitMap(nullptr, 0) {}
```

Comment:
GrowableBitMap(idx_t, bool) is dangerous to use because it calls reallocate, which calls the allocate function of the subclass. This works for ResourceBitMap, but not if you tried to use it for CHeapBitMap or ArenaBitMap. I'd suggest that we never let GrowableBitMap allocate memory in the constructor and instead push that responsiblity down into the subclasses.

Reply:
agree. I see your patch has changed this. This is safer. 
```
  GrowableBitMap() : GrowableBitMap(nullptr, 0) {}
  GrowableBitMap(bm_word_t* map, idx_t size_in_bits) : BitMap(map, size_in_bits) {}
```